File: apps/ios/Packages/MapleAPI/Sources/MapleWidgetData/WidgetSummary.swift
import Foundation

/// `GET /v2/widget_summary`, as this module sees it.
///
/// One request covering both Home Screen widgets, in place of the four the app
/// used to compose (`/v2/error_issues`, `/v2/services`, and two
/// `/v2/traces/timeseries`). It lives here rather than in `MapleAPI` because
/// both readers need it and only one of them can link a generated client: the
/// app fetches through `MapleAPI`, and the widget extension — which holds no
/// Clerk session and no 30k-line client — decodes the same shape by hand.
///
/// Plain `Codable` with explicit coding keys rather than a `keyDecodingStrategy`:
/// the strategy would be set by whoever happens to own the decoder, and this
/// payload is decoded in two processes.
public struct WidgetSummaryPayload: Codable, Sendable, Equatable {
	/// The wire's own version, independent of the API version. A payload from a
	/// newer server whose fields have changed meaning is rejected rather than
	/// rendered — see `isSupported`.
	public var schemaVersion: Int
	/// When the **server** read the data, not when the client received it. Every
	/// "updated 4m ago" on the Home Screen counts from here.
	public var generatedAt: Date
	/// Echoed by the server so a caller can prove the payload belongs to the
	/// organization it asked for before overwriting that organization's cached
	/// snapshot. There is deliberately no name: names come from the app's own
	/// membership index, and a second source could put one organization's name
	/// over another's numbers.
	public var organizationId: String
	/// The deployment environment the payload was filtered to, or nil for all of
	/// them.
	///
	/// Echoed for the same reason `organizationId` is, and it matters more here:
	/// snapshots are stored per (organization, environment), so a response whose
	/// filter the server ignored would overwrite a production widget's numbers
	/// with organization-wide ones and look entirely plausible doing it.
	public var deploymentEnvironment: String?
	public var issues: Issues
	public var throughput: Throughput

	public struct Issues: Codable, Sendable, Equatable {
		public var windowSeconds: Int
		/// More ongoing issues exist than `data` carries, so a count derived from
		/// it is a floor. The widget renders that as "20+".
		public var hasMore: Bool
		public var data: [Issue]

		public init(windowSeconds: Int, hasMore: Bool, data: [Issue]) {
			self.windowSeconds = windowSeconds
			self.hasMore = hasMore
			self.data = data
		}

		private enum CodingKeys: String, CodingKey {
			case windowSeconds = "window_seconds"
			case hasMore = "has_more"
			case data
		}
	}

	/// One issue, carrying the *raw* naming fields rather than a rendered title.
	/// The fallback between them is `WidgetIssueTitle`, which the app's issue
	/// list uses too — a title that resolved differently in the two places would
	/// read as two different issues.
	public struct Issue: Codable, Sendable, Equatable {
		public var id: String
		public var exceptionType: String
		public var errorLabel: String
		public var exceptionMessage: String
		public var serviceName: String
		/// Null for an untriaged issue, and also for a severity this build does
		/// not know: an unknown value decodes to nil rather than crashing a
		/// widget that cannot be updated without an App Store release.
		public var severity: String?
		public var occurrenceCount: Double
		public var lastSeenAt: Date
		public var isRegressed: Bool
		public var hasOpenIncident: Bool

		public init(
			id: String,
			exceptionType: String,
			errorLabel: String,
			exceptionMessage: String,
			serviceName: String,
			severity: String?,
			occurrenceCount: Double,
			lastSeenAt: Date,
			isRegressed: Bool,
			hasOpenIncident: Bool
		) {
			self.id = id
			self.exceptionType = exceptionType
			self.errorLabel = errorLabel
			self.exceptionMessage = exceptionMessage
			self.serviceName = serviceName
			self.severity = severity
			self.occurrenceCount = occurrenceCount
			self.lastSeenAt = lastSeenAt
			self.isRegressed = isRegressed
			self.hasOpenIncident = hasOpenIncident
		}

		private enum CodingKeys: String, CodingKey {
			case id
			case exceptionType = "exception_type"
			case errorLabel = "error_label"
			case exceptionMessage = "exception_message"
			case serviceName = "service_name"
			case severity
			case occurrenceCount = "occurrence_count"
			case lastSeenAt = "last_seen_at"
			case isRegressed = "is_regressed"
			case hasOpenIncident = "has_open_incident"
		}
	}

	public struct Throughput: Codable, Sendable, Equatable {
		public var windowSeconds: Int
		/// The bucket length behind every `points` array. Null when no series
		/// could be read, which is the signal to render the scalars without a
		/// sparkline rather than guess a unit.
		public var bucketSeconds: Int?
		public var services: [Service]
		/// The ungrouped organization series, in the same bucket counts as
		/// `services[].points`. Not the sum of those: the per-service series is
		/// capped at the charted few, so summing it would under-report a large
		/// organization's shape.
		public var totalPoints: [Double]

		public init(windowSeconds: Int, bucketSeconds: Int?, services: [Service], totalPoints: [Double]) {
			self.windowSeconds = windowSeconds
			self.bucketSeconds = bucketSeconds
			self.services = services
			self.totalPoints = totalPoints
		}

		private enum CodingKeys: String, CodingKey {
			case windowSeconds = "window_seconds"
			case bucketSeconds = "bucket_seconds"
			case services
			case totalPoints = "total_points"
		}
	}

	public struct Service: Codable, Sendable, Equatable {
		public var name: String
		public var throughputPerSecond: Double
		/// 0–1, not a percentage.
		public var errorRate: Double
		public var p95LatencyMs: Double
		/// Span **counts** per bucket, oldest first. Divided by `bucketSeconds`
		/// on the way into a snapshot so the sparkline and the headline provably
		/// carry the same unit.
		public var points: [Double]

		public init(
			name: String,
			throughputPerSecond: Double,
			errorRate: Double,
			p95LatencyMs: Double,
			points: [Double]
		) {
			self.name = name
			self.throughputPerSecond = throughputPerSecond
			self.errorRate = errorRate
			self.p95LatencyMs = p95LatencyMs
			self.points = points
		}

		private enum CodingKeys: String, CodingKey {
			case name
			case throughputPerSecond = "throughput_per_second"
			case errorRate = "error_rate"
			case p95LatencyMs = "p95_latency_ms"
			case points
		}
	}

	public init(
		schemaVersion: Int,
		generatedAt: Date,
		organizationId: String,
		deploymentEnvironment: String? = nil,
		issues: Issues,
		throughput: Throughput
	) {
		self.schemaVersion = schemaVersion
		self.generatedAt = generatedAt
		self.organizationId = organizationId
		self.deploymentEnvironment = deploymentEnvironment
		self.issues = issues
		self.throughput = throughput
	}

	private enum CodingKeys: String, CodingKey {
		case schemaVersion = "schema_version"
		case generatedAt = "generated_at"
		case organizationId = "organization_id"
		case deploymentEnvironment = "deployment_environment"
		case issues
		case throughput
	}

	/// The version this build was written against. A payload above it may have
	/// changed what an existing field *means*, which is the one thing a
	/// tolerant decoder cannot absorb.
	public static let supportedSchemaVersion = 1

	public var isSupported: Bool { schemaVersion <= Self.supportedSchemaVersion }
}

/// How an issue names itself, from the raw contract fields.
///
/// Shared rather than private to either reader: the app's issue list and the
/// Home Screen widget show the same rows, and a title that falls back
/// differently between them reads as two different issues.
public enum WidgetIssueTitle {
	/// The exception type, or — for the kinds that carry none (integration and
	/// alert issues) — the label, or failing that the message.
	public static func title(exceptionType: String, errorLabel: String, exceptionMessage: String) -> String {
		let type = exceptionType.trimmingCharacters(in: .whitespacesAndNewlines)
		if !type.isEmpty { return type }
		let label = errorLabel.trimmingCharacters(in: .whitespacesAndNewlines)
		return label.isEmpty ? exceptionMessage : label
	}

	/// The message, suppressed when it would merely restate the title — which is
	/// what happens once the title has fallen back to the label.
	public static func subtitle(
		exceptionType: String,
		errorLabel: String,
		exceptionMessage: String
	) -> String? {
		let message = exceptionMessage.trimmingCharacters(in: .whitespacesAndNewlines)
		let title = title(
			exceptionType: exceptionType,
			errorLabel: errorLabel,
			exceptionMessage: exceptionMessage
		)
		guard !message.isEmpty, !message.hasPrefix(title), !title.hasPrefix(message) else { return nil }
		return message
	}
}

extension WidgetSummaryPayload {
	/// The issues snapshot this payload describes.
	///
	/// `IssuesSnapshot.make` still does the ranking, truncation, and counting:
	/// the server returns a wider page than the widget draws precisely so that
	/// `openCount` means something, and one place has to reconcile the headline
	/// with the rows.
	///
	/// - Parameter organizationName: from the app's membership index, never from
	///   the payload. The index is corrected the moment memberships load,
	///   whereas a name baked into a snapshot is only as current as the round
	///   that wrote it.
	public func issuesSnapshot(organizationName: String?) -> IssuesSnapshot {
		IssuesSnapshot.make(
			organizationId: organizationId,
			organizationName: organizationName,
			generatedAt: generatedAt,
			issues: issues.data.map { issue in
				WidgetIssue(
					id: issue.id,
					title: WidgetIssueTitle.title(
						exceptionType: issue.exceptionType,
						errorLabel: issue.errorLabel,
						exceptionMessage: issue.exceptionMessage
					),
					subtitle: WidgetIssueTitle.subtitle(
						exceptionType: issue.exceptionType,
						errorLabel: issue.errorLabel,
						exceptionMessage: issue.exceptionMessage
					),
					serviceName: issue.serviceName,
					severity: issue.severity.flatMap(WidgetIssueSeverity.init(rawValue:)),
					occurrenceCount: issue.occurrenceCount,
					lastSeenAt: issue.lastSeenAt,
					isRegressed: issue.isRegressed,
					hasOpenIncident: issue.hasOpenIncident
				)
			},
			hasMore: issues.hasMore
		)
	}

	/// The throughput snapshot this payload describes.
	///
	/// The organization total's *numbers* are summed from every service row —
	/// which is why the server returns more of them than the widget charts —
	/// while only its *shape* comes from the ungrouped series, the one thing
	/// the rows cannot provide.
	public func throughputSnapshot() -> ThroughputSnapshot {
		let services = throughput.services.map { service in
			ServiceThroughput(
				name: service.name,
				throughputPerSecond: service.throughputPerSecond,
				errorRate: service.errorRate,
				p95LatencyMs: service.p95LatencyMs,
				points: perSecond(service.points)
			)
		}
		var overall = ServiceThroughput.total(of: services)
		let total = perSecond(throughput.totalPoints)
		if !total.isEmpty { overall.points = total }
		return ThroughputSnapshot.make(
			organizationId: organizationId,
			generatedAt: generatedAt,
			windowMinutes: throughput.windowSeconds / 60,
			services: services,
			overall: overall
		)
	}

	/// Spans per bucket → spans per second, so the sparkline carries the same
	/// unit as the headline. A missing or nonsensical bucket length leaves the
	/// series out rather than drawing counts as if they were rates.
	private func perSecond(_ values: [Double]) -> [Double] {
		guard let bucketSeconds = throughput.bucketSeconds, bucketSeconds > 0 else { return [] }
		return values.map { $0 / Double(bucketSeconds) }
	}
}

File: packages/domain/src/http/v2/widget-summary.ts
import { HttpApiEndpoint, HttpApiGroup, OpenApi } from "effect/unstable/httpapi"
import { Schema } from "effect"
import { ErrorPersistenceError, IssueSeverity } from "../errors"
import { OrgId, ServiceName } from "../../primitives"
import { AuthorizationV2 } from "./auth"
import { wireExample, Timestamp } from "./envelopes"
import { publicError } from "./public-error"
import { V2QueryErrors } from "./query-errors"
import { ErrorIssuePublicId } from "./resource-ids"

/**
 * The iOS Home Screen widgets' one read.
 *
 * A separate resource family rather than a shaped view over `/v2/error_issues`
 * + `/v2/services` + `/v2/traces/timeseries`, for two reasons that are not
 * "three requests is more than one":
 *
 * 1. **It is the credential's fence.** `requiredScopeForRoute` derives an API
 *    key's required scope from the first path segment, so a key scoped
 *    `widget_summary:read` can reach exactly this endpoint and nothing else.
 *    Composed from the generic endpoints, the same widget would need
 *    `error_issues:read` + `services:read` + `traces:read` — an organization
 *    read key, sitting on a phone. For the same reason this must never be
 *    nested under another family's prefix.
 * 2. **A widget extension has no budget to compose.** It is woken by WidgetKit
 *    with seconds of wall clock and tens of megabytes, and it deliberately does
 *    not link the generated client. One request it can decode by hand is the
 *    only shape that fits.
 *
 * The windows are the server's, not the caller's: "ongoing issues" and "traffic
 * right now" are product definitions the widgets render, and a query parameter
 * would let two builds of the app disagree about what the Home Screen means.
 *
 * `deployment_environment` is a parameter anyway, and the two are not in
 * tension. A window is what the Home Screen *means*; an environment is which
 * slice of the organization the person holding the phone asked to look at. The
 * server cannot know the second, and two builds cannot disagree about it —
 * whatever they send comes back echoed on the response.
 */

/** Ongoing means the app's "Needs attention" filter over the day Home considers recent. */
export const WIDGET_SUMMARY_ISSUES_WINDOW_SECONDS = 60 * 60 * 24
/** Throughput is a "right now" number — Home's rate window, and the Services tab's. */
export const WIDGET_SUMMARY_THROUGHPUT_WINDOW_SECONDS = 60 * 60

/**
 * Enough issues that `open_count` is meaningful and the rows shown are really
 * the worst ones; past this the widget renders a floor ("20+").
 */
export const WIDGET_SUMMARY_ISSUE_LIMIT = 20
/**
 * The widget charts far fewer, but the organization total is summed across
 * every service — so this is deliberately wider than what is drawn.
 */
export const WIDGET_SUMMARY_SERVICE_LIMIT = 50
/** Series returned with per-service buckets. Matches what the widget can draw. */
export const WIDGET_SUMMARY_SERIES_LIMIT = 12

/**
 * The wire's own version, independent of the API version.
 *
 * The reader is an App Store binary that cannot be updated in step with a
 * deploy, so the server needs a way to know which shape a caller understands
 * before it adds a field that changes how one is drawn. Bump only when the
 * meaning of an existing field changes; new optional fields do not need it.
 */
export const WIDGET_SUMMARY_SCHEMA_VERSION = 1

export const V2WidgetSummaryIssue = Schema.Struct({
	id: ErrorIssuePublicId,
	/**
	 * The raw naming fields, not a rendered title. The app's issue list and the
	 * widget must fall back identically — a title that resolves differently in
	 * the two places reads as two different issues — and the one implementation
	 * of that fallback lives on the client, next to the list that also uses it.
	 */
	exception_type: Schema.String,
	error_label: Schema.String,
	exception_message: Schema.String,
	service_name: Schema.String,
	severity: Schema.NullOr(IssueSeverity),
	occurrence_count: Schema.Number,
	last_seen_at: Timestamp,
	/** Fixed, then seen again — the state that means a deploy undid a fix. */
	is_regressed: Schema.Boolean,
	has_open_incident: Schema.Boolean,
}).annotate({
	identifier: "WidgetSummaryIssue",
	title: "Widget summary issue",
	description: "One ongoing error issue, reduced to what a Home Screen row can render.",
})
export type V2WidgetSummaryIssue = Schema.Schema.Type<typeof V2WidgetSummaryIssue>

export const V2WidgetSummaryService = Schema.Struct({
	name: ServiceName,
	throughput_per_second: Schema.Number,
	/** 0–1, not a percentage. */
	error_rate: Schema.Number,
	p95_latency_ms: Schema.Number,
	/**
	 * Span counts per bucket, oldest first — **counts, not rates**. The client
	 * divides by `bucket_seconds` so the sparkline and the headline provably
	 * carry the same unit, and so a bucket length the client cannot make sense
	 * of drops the series rather than drawing counts as if they were rates.
	 */
	points: Schema.Array(Schema.Number),
}).annotate({
	identifier: "WidgetSummaryService",
	title: "Widget summary service",
	description: "One service's traffic over the throughput window.",
})
export type V2WidgetSummaryService = Schema.Schema.Type<typeof V2WidgetSummaryService>

export const V2WidgetSummaryIssues = Schema.Struct({
	window_seconds: Schema.Number,
	/**
	 * More ongoing issues exist than `data` carries, so a count derived from it
	 * is a floor. The widget renders that as "20+" rather than a wrong total.
	 */
	has_more: Schema.Boolean,
	data: Schema.Array(V2WidgetSummaryIssue),
}).annotate({
	identifier: "WidgetSummaryIssues",
	title: "Widget summary issues",
})

export const V2WidgetSummaryThroughput = Schema.Struct({
	window_seconds: Schema.Number,
	/**
	 * Bucket length behind every `points` array. Null when no series could be
	 * read, which tells the client to render the scalars without a sparkline
	 * instead of guessing a unit.
	 */
	bucket_seconds: Schema.NullOr(Schema.Number),
	services: Schema.Array(V2WidgetSummaryService),
	/**
	 * The ungrouped organization series, in the same bucket counts as
	 * `services[].points`.
	 *
	 * Not the sum of `services[].points`: the per-service series is capped at
	 * the charted few, so summing it would under-report a large organization's
	 * shape. The scalar total is still derived client-side from every service
	 * row, which is why only the series appears here.
	 */
	total_points: Schema.Array(Schema.Number),
}).annotate({
	identifier: "WidgetSummaryThroughput",
	title: "Widget summary throughput",
})

export const V2WidgetSummary = Schema.Struct({
	object: Schema.Literal("widget_summary").annotate({
		description: 'The object type — always `"widget_summary"`.',
	}),
	schema_version: Schema.Number.annotate({
		description:
			"The widget wire shape's own version. Clients that do not recognise it should keep rendering their last good snapshot rather than decode against a shape whose fields may have changed meaning.",
	}),
	/** When the server read the data — the age every widget renders from. */
	generated_at: Timestamp,
	/**
	 * Echoed so the caller can prove the payload belongs to the organization it
	 * asked for before writing it over that organization's cached snapshot.
	 * There is deliberately no name here: the client already resolves names from
	 * its own membership index, and a second source would let a widget render
	 * one organization's name over another's numbers.
	 */
	organization_id: OrgId,
	/**
	 * The environment the payload was filtered to, or null for all of them.
	 *
	 * Echoed for the same reason `organization_id` is: the caller keeps one
	 * cached snapshot per (organization, environment) and has to prove a
	 * response belongs to the slot it is about to overwrite. Without it a
	 * request whose filter the server ignored would quietly overwrite a
	 * production widget's numbers with organization-wide ones.
	 */
	deployment_environment: Schema.NullOr(Schema.String),
	issues: V2WidgetSummaryIssues,
	throughput: V2WidgetSummaryThroughput,
}).annotate({
	identifier: "WidgetSummary",
	title: "Widget summary",
	description:
		"Everything the Maple iOS Home Screen widgets draw, in one response: ongoing error issues over the last day, and per-service traffic over the last hour.",
	examples: [
		wireExample({
			object: "widget_summary",
			schema_version: 1,
			generated_at: "2026-08-21T09:10:00.000Z",
			organization_id: "org_2abcDEF",
			deployment_environment: "production",
			issues: {
				window_seconds: 86_400,
				has_more: false,
				data: [
					{
						id: "iss_YofPTrK9782DWwcnXhpcCw",
						exception_type: "TypeError",
						error_label: "checkout",
						exception_message: "Cannot read properties of undefined",
						service_name: "api",
						severity: "critical",
						occurrence_count: 412,
						last_seen_at: "2026-08-21T09:08:12.000Z",
						is_regressed: false,
						has_open_incident: true,
					},
				],
			},
			throughput: {
				window_seconds: 3600,
				bucket_seconds: 300,
				services: [
					{
						name: "api",
						throughput_per_second: 12.5,
						error_rate: 0.012,
						p95_latency_ms: 184,
						points: [3600, 3720, 3540],
					},
				],
				total_points: [5200, 5310, 5180],
			},
		}),
	],
})
export type V2WidgetSummary = Schema.Schema.Type<typeof V2WidgetSummary>

export const V2WidgetSummaryQuery = Schema.Struct({
	/**
	 * Restrict every half of the payload to one deployment environment. Absent
	 * means the whole organization, which is what widgets placed before this
	 * parameter existed keep asking for.
	 */
	deployment_environment: Schema.optional(Schema.String),
}).annotate({ identifier: "WidgetSummaryQuery", title: "Widget summary query" })

export class V2WidgetSummaryApiGroup extends HttpApiGroup.make("widgetSummary")
	.add(
		HttpApiEndpoint.get("retrieve", "/", {
			query: V2WidgetSummaryQuery,
			success: V2WidgetSummary,
			error: [publicError(ErrorPersistenceError), ...V2QueryErrors],
		}).annotateMerge(
			OpenApi.annotations({
				identifier: "getWidgetSummary",
				summary: "Retrieve the mobile widget summary",
				description:
					"Returns ongoing error issues and per-service traffic in a single small payload sized for a Home Screen widget. The windows are fixed by the server; `deployment_environment` optionally narrows both halves to one environment and is echoed on the response. Requires the `widget_summary:read` scope.",
			}),
		),
	)
	.prefix("/v2/widget_summary")
	.middleware(AuthorizationV2)
	.annotateMerge(
		OpenApi.annotations({
			title: "Widget Summary",
			description:
				"The single read behind the Maple mobile Home Screen widgets. Deliberately its own scope family so a device credential can be fenced to it alone.",
		}),
	) {}
